File: bin/prettyprompt.cpp
#include <QtCore/QCoreApplication>
#include <QtCore/QDir>
#include <QtCore/QProcess>

#include <cstdio>
#include <cstdlib>
#include <memory>
#include <unistd.h>

QByteArray readFile(const QString& path)
{
	QFile file(path);
	if (file.open(QIODevice::ReadOnly))
		return file.readAll();
	else
		return QByteArray();
}

class RepoReader
{
	protected:
		bool m_isRepo;
		QStringList m_basePathParts, m_pathPartsInRepo;
	public:
		RepoReader(const QStringList& pathParts) : m_isRepo(false), m_basePathParts(pathParts) {}
		virtual ~RepoReader() {}

		bool isRepo() const { return m_isRepo; }
		QByteArray printablePath() const
		{
			if (m_isRepo && !m_pathPartsInRepo.isEmpty())
			{
				const QByteArray basePath = (m_basePathParts.join(QChar('/')) + QChar('/')).toLocal8Bit();
				const QByteArray pathInRepo = m_pathPartsInRepo.join(QChar('/')).toLocal8Bit();
				return "\e[0;36m" + basePath + "\e[1;36m" + pathInRepo + "\e[0m";
			}
			else
			{
				const QByteArray path = (m_basePathParts + m_pathPartsInRepo).join(QChar('/')).toLocal8Bit();
				return "\e[1;36m" + path + "\e[0m";
			}
		}
		virtual QByteArray extraInfo() const
		{
			return QByteArray();
		}
};

class GitRepoReader : public RepoReader
{
	private:
		QByteArray m_branch, m_commit;
	public:
		GitRepoReader(const QStringList& pathParts) : RepoReader(pathParts)
		{
			//NOTE: path is in m_basePathParts at the beginning
			QString gitDirPath;
			while (!m_basePathParts.isEmpty())
			{
				gitDirPath = m_basePathParts.join(QChar('/')) + QLatin1String("/.git");
				//find Git repo for current directory
				if (QDir(gitDirPath).exists())
				{
					m_isRepo = true;
					break;
				}
				//ascend in directory hierarchy as necessary
				m_pathPartsInRepo.prepend(m_basePathParts.takeLast());
			}
			if (m_isRepo)
			{
				//determine current branch and commit
				QByteArray headRef = readFile(gitDirPath + QLatin1String("/HEAD")).simplified();
				if (headRef.startsWith("ref: refs/"))
				{
					const QByteArray refSpec = headRef.mid(5);
					QByteArray headRef2 = headRef.mid(10);
					if (headRef2.startsWith("heads/"))
						//current HEAD is a branch
						m_branch = headRef2.mid(6);
					else
						//current HEAD is a remote or tag (include type specification in branch name)
						m_branch = headRef2;
					//read corresponding file to find commit
					m_commit = readFile(gitDirPath + QChar('/') + refSpec).simplified();
					if (m_commit.isEmpty())
					{
						const QList<QByteArray> packedRefs = readFile(gitDirPath + QLatin1String("/packed-refs")).split('\n');
						foreach (const QByteArray& packedRef, packedRefs)
							if (packedRef.endsWith(refSpec))
							{
								m_commit = packedRef.mid(0, 40);
								break;
							}
					}
				}
				else
				{
					//current HEAD is a commit
					m_branch = QByteArray();
					m_commit = headRef;
				}
			}
		}
		virtual QByteArray extraInfo() const
		{
			if (!m_isRepo)
				return QByteArray();
			QByteArray result;
			//branch
			if (m_branch.isEmpty())
				result += " on \e[1;41mno branch\e[0m";
			else
				result += " on branch " + m_branch;
			//commit
			if (!m_commit.isEmpty())
				result += " at " + m_commit.mid(0, 7);
			//done
			return result;
		}
};

class SvnRepoReader : public RepoReader
{
	private:
		int m_revision;
	public:
		SvnRepoReader(const QStringList& pathParts) : RepoReader(pathParts), m_revision(-1)
		{
			//find highest .svn directory
			//NOTE: path is in m_basePathParts at the beginning
			QString svnDirPath = m_basePathParts.join(QChar('/')) + QLatin1String("/.svn");;
			while (QDir(svnDirPath).exists())
			{
				m_pathPartsInRepo.prepend(m_basePathParts.takeLast());
				svnDirPath = m_basePathParts.join(QChar('/')) + QLatin1String("/.svn");
			}
			m_isRepo = !m_pathPartsInRepo.isEmpty();
			if (m_isRepo)
			{
				//correct a systematic error: the above loop moves one path part too much
				m_basePathParts.append(m_pathPartsInRepo.takeFirst());
				m_isRepo = true;
			}
			//find revision
			QProcess svnInfo;
			svnInfo.setEnvironment(QProcess::systemEnvironment() << "LANG=C");
			svnInfo.start("svn", QStringList() << "info");
			svnInfo.waitForFinished();
			QList<QByteArray> svnInfoLines = svnInfo.readAllStandardOutput().split('\n');
			foreach (const QByteArray& svnInfoLine, svnInfoLines)
				if (svnInfoLine.startsWith("Revision: "))
					m_revision = svnInfoLine.mid(10).toInt();
		}
		virtual QByteArray extraInfo() const
		{
			if (!m_isRepo || m_revision < 0)
				return QByteArray();
			else
				return " at revision " + QByteArray::number(m_revision);
		}
};

static int atoi(const QByteArray& str, int defaultValue)
{
	bool ok;
	const int result = str.toInt(&ok);
	return ok ? result : defaultValue;
}

int main(int argc, char** argv)
{
	QCoreApplication app(argc, argv);
	//get basic information: username, hostname, cwd, shell name
	const int bufferLength = 1024;
	char buffer[bufferLength];
	gethostname(buffer, 1024);
	QByteArray hostname(buffer);
	const QByteArray username(getenv("LOGNAME")), term(getenv("TERM"));
	const QStringList cwdParts = QDir::current().absolutePath().split(QChar('/'));
	const QByteArray shellName(getenv("PRETTYPROMPT_SHELL"));
	int shellLevel = atoi(getenv("SHLVL"), 1);
#if 0
	if (shellName == "zsh")
		shellLevel += 1; //work around a counting problem in zsh
#endif
	shellLevel = qMax(1, shellLevel);
	//determine appearance of username@hostname and prompt symbol
	const char* usernameColor = "0";
	if (username == "root")
		usernameColor = "0;1;41";
	const char* hostnameColor = "0;33";
	if (hostname == "magrathea" || hostname == "magrathea.local")
	{
		hostname = "magrathea";
		hostnameColor = "0;32";
	}
	else if (hostname == "maximegalon")
		hostnameColor = "0;31";
	else if (hostname == "vserver3190")
		hostname = "bethselamin.de";
	//find repos etc.
	RepoReader* repoReader = new GitRepoReader(cwdParts);
	if (!repoReader->isRepo())
	{
		delete repoReader;
		repoReader = new SvnRepoReader(cwdParts);
	}
	//print "username@hostname cwd"
	printf(
		 "\e[%sm%s\e[0m"   //username
		"@\e[%sm%s\e[0m"   //hostname
		"-%s"              //terminal (esp. for identifying screen)
		" %s%s\n"          //cwd, extra info from repo reader
		, usernameColor, username.data()
		, hostnameColor, hostname.data()
		, term.data()
		, repoReader->printablePath().data(), repoReader->extraInfo().data()
	);
	//print shell name and shell level
	printf("%s%i$ ", shellName.data(), shellLevel);
	//cleanup
	delete repoReader;
	return 0;
}

File: bin/prettyprompt.py
#!/usr/bin/python

import os, os.path as op, pwd, re, socket, sys, subprocess as sp

def catFile(file):
    """ Returns the content of the given file. No exception handling. """
    try:
        return "".join(open(file).readlines())
    except IOError:
        return ""

def colored(string, color):
    """ Wraps a string in terminal escape sequences to activate given color. """
    return "\033[%sm%s\033[0m" % (color, string)

class NotARepoException(Exception):
    pass

def recognizeGitRepo(path):
    """ Returns a triple of repo path, path in repo, and repo status.
        E.g. ("/foo", "bar", "on branch master at 631d7a2") for
        path == "/foo/bar" and os.path.exists("/foo/.git").

        Throws NotARepoException if .git cannot be found.
    """
    basePath = op.realpath(path)
    subPath = ""
    # find Git repo
    while not op.exists(op.join(basePath, ".git")):
        # ascend in directory hierarchy if possible
        basePath, newSubDir = op.split(basePath)
        subPath = op.join(newSubDir, subPath)
        # root directory reached? -> not in Git repo
        if newSubDir == "":
            raise NotARepoException

    # determine current branch and commit
    gitDir = op.join(basePath, ".git")
    headRef = catFile(op.join(gitDir, "HEAD")).strip()
    if headRef.startswith("ref: refs/"):
        refSpec = headRef[5:]
        headRef2 = headRef[10:]
        if headRef2.startswith("heads/"):
            # current HEAD is a branch
            branch = headRef2[6:]
        else:
            # current HEAD is a remote or tag -> include type specification
            branch = headRef2
        branchSpec = "branch " + branch
        # read corresponding file to find commit
        commit = catFile(op.join(gitDir, refSpec)).strip()
        if commit == "" and op.exists(op.join(gitDir, "packed-refs")):
            packedRefs = open(op.join(gitDir, "packed-refs")).readlines()
            packedRefs = map(str.strip, packedRefs)
            for packedRef in packedRefs:
                if packedRef.endswith(refSpec):
                    commit = packedRef[0:40]
                    break
    else:
        # current HEAD is detached
        branchSpec = colored("no branch", "1;41")
        commit = headRef

    if commit == "": # before initial commit
        branchSpec = branchSpec + colored(" before initial commit", "1;30")
        extraInfo = "on %s" % (branchSpec)
    else:
        extraInfo = "on %s at %s" % (branchSpec, commit[0:7])
    return basePath, subPath, extraInfo

def toInt(string, default):
    try:
        return int(string)
    except ValueError: # not a number literal
        return int(default)

# start writing, define shorthand for sys.stdout.write
stdoutBuffer = ""
def ssw(string):
    string = str(string)
    global stdoutBuffer
    stdoutBuffer += string
    sys.stdout.write(string)

# find username and hostname
username = pwd.getpwuid(os.getuid())[0]
hostname = socket.gethostname()
if hostname.endswith(".site"):
    hostname = hostname[:-5]
elif hostname.endswith(".local"):
    hostname = hostname[:-6]
elif hostname == "vserver3190":
    hostname = "bethselamin.de"

# select color for hostname, skip username display for my most common non-root user
commonUser    = os.environ.get("PRETTYPROMPT_COMMONUSER", "stefan")
hostnameColor = os.environ.get("PRETTYPROMPT_HOSTCOLOR",  "0;33")

# print username (with different color for some contexts)
if username != commonUser:
    usernameColor = { "root": "0;1;41" }.get(username, "0")
    ssw(colored(username, usernameColor) + "@")

# print hostname with different color for most hosts
ssw(colored(hostname, hostnameColor))

# print terminal (esp. for identifiying screen)
termName = os.environ["TERM"]
if termName == "xterm-256color":
    ssw(" ")
else:
    ssw("-" + termName + " ")

# find root build directory (used for displaying cwd inside build tree in condensed form)
buildRoot = os.environ["BUILD_ROOT"]
if buildRoot != "":
    buildRoot = op.realpath(buildRoot)

# method to shorten home directory to "~"
def stripHome(path):
    h = os.environ.get("HOME")
    hh = h + "/"
    if path.startswith(hh):
        return path[len(hh):]
    if path.startswith(h):
        return path[len(h):]
    else:
        return path

# find cwd, does it exist?
try:
    cwd = op.realpath(os.getcwd())
    cwdExists = True
except OSError:
    cwd = os.environ.get("PWD")
    cwdExists = False
    # find last existing parent directory and display similar to repo, but with red alert color
    basePath, subPath = cwd, ""
    while not op.exists(basePath):
        basePath, newSubDir = op.split(basePath)
        subPath = op.join(newSubDir, subPath)
    subPath = subPath.rstrip("/")
    if subPath == "":
        ssw(colored(stripHome(basePath), "1;36"))
    else:
        ssw(colored(stripHome(basePath) + "/", "1;36") + colored(subPath, "1;31"))
    ssw(" " + colored("could not stat cwd", "1;41"))

# is cwd in build tree? -> if so, print and process source dir instead
if buildRoot == "":
    isBuildDir = False
else:
    rel = op.relpath(cwd, buildRoot)
    isBuildDir = not rel.startswith("..")
    if isBuildDir:
        cwd = op.normpath("/" + rel)

# check if cwd is in a repo?
if cwdExists:
    isRepo = False

    try:
        repoBase, repoPath, repoStatus = recognizeGitRepo(cwd)
        isRepo = True
    except NotARepoException:
        pass

    # print cwd, builddir markers and repo status (if any)
    buildDirMarker = colored("BUILD", "1;35")
    if isBuildDir:
        ssw(buildDirMarker + " ")
    if not isRepo:
        ssw(colored(stripHome(cwd), "1;36"))
    else:
        if repoPath == "":
            ssw(colored(stripHome(repoBase), "1;36"))
        else:
            repoPath = repoPath.rstrip("/")
            ssw(colored(stripHome(repoBase) + "/", "0;36") + colored(repoPath, "1;36"))
        ssw(" " + repoStatus)

# display currently selected cloud (if any)
cloudKey = os.environ.get("CURRENT_OS_CLOUD", "")
if cloudKey != "":
    ssw(colored(" on cloud " + cloudKey, "0;37"))

# how many actual characters have been written?
colorSpecRx = "\033\\[[^m]*m"
printedStr = "".join(re.split(colorSpecRx, stdoutBuffer))
printedLen = len(printedStr)
if not printedStr.endswith(" "):
    ssw(" ")
    printedLen += 1
# draw separation line using terminal width
try:
    termWidth = int(sys.argv[1])
except:
    termWidth = 0
if printedLen < termWidth:
    ssw(colored("-" * (termWidth - printedLen), "1"))

# final prompt: shell name and shell level
shellName = os.environ["PRETTYPROMPT_SHELL"]
shellLevel = toInt(os.environ["SHLVL"], 1)
if "zsh" in shellName:
    shellLevel += 1
shortenedShellName = { "zsh": "Z", "bash": "B" }.get(shellName, shellName)

ssw("\n%s%i$ " % (shortenedShellName, shellLevel))

# DEBUG
ssw("\n")
